File: android/app/src/main/kotlin/io/github/benderblog/cmcalc/CalculatorBridge.kt
// Copyright 2025 BenderBlog Rodriguez
// SPDX-License-Identifier: GPL-2.0-or-later

package io.github.benderblog.cmcalc

import CalcResult
import CalculatorWrapper
import com.jherkenhoff.libqalculate.Calculator
import com.jherkenhoff.libqalculate.MessageType

// Suppose to be "always running"
class CalculatorBridge(val calc: Calculator) : CalculatorWrapper {
    init {
        calc.loadGlobalDefinitions()
    }

    private fun parseCommand(command: String) : String {
        val toParse = calc.parse(command)
        toParse.format()
        return toParse.print()
    }

    override fun calculate(command: String, callback: (Result<CalcResult>) -> Unit) {
        calc.clearMessages()

        val result = calc.calculate(calc.unlocalizeExpression(command))
        val message = calc.message()

        val resultMsg : CalcResult = CalcResult(
            resultType = if (message == null) {
                ResultType.SUCCESS
            } else {
                when (message.type()) {
                    MessageType.MESSAGE_ERROR -> ResultType.FAILURE
                    MessageType.MESSAGE_WARNING ->ResultType.WARNING
                    else -> ResultType.SUCCESS
                }
            },
            message = if (message == null) {
                ""
            } else {
                message.message()
            },
            parsed = parseCommand(command),
            result = result.print(),
        )
        callback(Result.success(resultMsg))
    }

    override fun parseCommand(command: String, callback: (Result<String>) -> Unit) {
        callback(Result.success(parseCommand(command)))
    }
}

File: pigeon/calculator_bridge.dart
// Copyright 2025 BenderBlog Rodriguez
// SPDX-License-Identifier: GPL-2.0-or-later

// A general package for Qalculate!

import 'package:pigeon/pigeon.dart';

@ConfigurePigeon(PigeonOptions(
  dartOut: 'lib/calculator_bridge/calculator_bridge.g.dart',
  dartOptions: DartOptions(),
  swiftOut: 'ios/Runner/calculator_bridge.g.swift',
  swiftOptions: SwiftOptions(),
  kotlinOut: 'android/app/src/main/kotlin/io/github/benderblog/cmcalc/CalculatorBridge.g.kt',
  kotlinOptions: KotlinOptions(),
))

enum ResultType {
  success,
  warning,
  failure,
}

class CalcResult {
  CalcResult({
    required this.resultType,
    required this.message,
    required this.parsed,
    required this.result,
  });
  ResultType resultType;
  String message;
  String parsed;
  String result;
}


@HostApi()
abstract class CalculatorWrapper {
  @async
  CalcResult calculate(String command);

  @async
  String parseCommand(String command);
}